docs/programmers_guide/witness_formal_spec.md

```diff
@@ -164,6 +165,125 @@ The exact implementation details are undefined in this spec.
 
 ## Instructions
 
+### `LEAF key raw_value`
+
+**Substitution rules**
+
+```
+LEAF(key, raw_value) |=> 
+STACK(ShortNode{key, ValueNode(raw_value)}, COMPACT_KECCAK(RLP(RLP(raw_value))))
+```
+
+### `EXTENSION key`
+
+**Substitution rules**
+
+```
+GUARD node != nil
+
+STACK(node, hash) EXTENSION(key) |=>
+STACK(ShortNode{key, node}, hash)
+
+---
+
+GUARD node == nil
+
+STACK(nil, hash) EXTENSION(key) |=>
+STACK(ShortNode{key, HashNode(hash)}, hash)
+```
+
+### `HASH raw_hash`
+
+Pushes a `HashNode` to stack.
+
+**Substitution rules**
+
+```
+HASH(raw_hash) |=>
+STACK(HashNode{raw_hash}, raw_hash)
+```
+
+### `CODE raw_code`
+
+Pushes an nil node + the code hash to the stack.
+
+```
+CODE(raw_code) |=>
+STACK(nil, RLP(KECCAK(raw_code)))
+```
+
+### `ACCOUNT_LEAF key nonce balance has_code has_storage`
+
+**Substitution rules**
+
+```
+
+GUARD has_code == true
+GUARD storage_node == nil
+GUARD has_storage == true
+
+STACK(nil, code_hash) STACK(nil, storage_hash) ACCOUNT_LEAF(key, nonce, balance, has_code, has_storage) |=>
+STACK(ShortNode{key, AccountNode{Account(nonce, balance, HashNode{storage_hash}, storage_hash, code_hash)}}, ACC_HASH(key, account))
+--
+
+GUARD has_code == true
+GUARD storage_node != nil
+GUARD has_storage == true
+
+STACK(nil, code_hash) STACK(storage_node, storage_hash) ACCOUNT_LEAF(key, nonce, balance, has_code, has_storage) |=>
+STACK(ShortNode{key, AccountNode{Account(nonce, balance, storage_node, storage_hash, code_hash)}}, ACC_HASH(key, account))
+
+--
+
+GUARD has_code == false
+GUARD storage_node != nil
+GUARD has_storage == true
+
+STACK(storage_node, storage_hash) ACCOUNT_LEAF(key, nonce, balance, has_code, has_storage) |=>
+STACK(ShortNode{key, AccountNode{Account(nonce, balance, storage_node, storage_hash, nil)}}, ACC_HASH(key, account))
+
+---
+
+GUARD has_code == false
+GUARD storage_node == nil
+GUARD has_storage == true
+
+STACK(nil, storage_hash) ACCOUNT_LEAF(key, nonce, balance, has_code, has_storage) |=>
+STACK(ShortNode{key, AccountNode{Account(nonce, balance, storage_node, storage_hash, nil)}}, ACC_HASH(key, account))
+
+--
+
+GUARD has_code == true
+GUARD has_storage == false
+
+STACK(nil, code_hash) ACCOUNT_LEAF(key, nonce, balance, has_code, has_storage) |=>
+STACK(ShortNode{key, AccountNode(Account(nonce, balance, nil, nil, code_hash)}}, ACC_HASH(key, account))
+
+--
+
+GUARD has_code == false
+GUARD has_storage == false
+
+ACCOUNT_LEAF(key, nonce, balance, has_code, has_storage) |=>
+STACK(ShortNode{key, AccountNode{Account(nonce, balance, nil, nil, nil)}}, ACC_HASH(key, account))
+
+```
+
+### `EMPTY_ROOT`
+
+Pushes an empty node + an empty hash to the stack
+
+**Substitution rules**
+
+```
+EMPTY_ROOT |=>
```

Comment:
Did we end up using empty root or not?

Reply:
We use it for empty storage tries, afaik. But I can double-check.

Reply:
yeah, in [witness_builder.go:L180](https://github.com/ledgerwatch/turbo-geth/blob/master/trie/witness_builder.go#L180)